# Lab 2 Data Analysis of U.S. Electric Vehicle Data

## Data Source

This data is sourced from: [https://catalog.data.gov/dataset/electric-vehicle-population-data]

# View Data

In [1]:
import warnings
warnings.simplefilter('ignore')

import pandas as pd

df = pd.read_csv('s3://ia-340-lab-1-bucket/Electric_Vehicle_Population_Data.csv') # load the data from my bucket
df.head() #show the top 5 rows

VIN (1-10)     County          City State  Postal Code  Model Year    Make  \
0  5YJ3E1EB5K     Yakima        Yakima    WA      98901.0        2019   TESLA   
1  1C4RJXU67R     Kitsap  Port Orchard    WA      98367.0        2024    JEEP   
2  KNDCD3LD0N  Snohomish      Lynnwood    WA      98036.0        2022     KIA   
3  5UXKT0C37H       King        Auburn    WA      98001.0        2017     BMW   
4  1N4AZ0CP1D     Skagit  Mount Vernon    WA      98273.0        2013  NISSAN   

      Model                   Electric Vehicle Type  \
0   MODEL 3          Battery Electric Vehicle (BEV)   
1  WRANGLER  Plug-in Hybrid Electric Vehicle (PHEV)   
2      NIRO  Plug-in Hybrid Electric Vehicle (PHEV)   
3        X5  Plug-in Hybrid Electric Vehicle (PHEV)   
4      LEAF          Battery Electric Vehicle (BEV)   

  Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0           Clean Alternative Fuel Vehicle Eligible           220.0   
1             Not eligible due to low battery range            21.0   
2             Not eligible due to low battery range            26.0   
3             Not eligible due to low battery range            14.0   
4           Clean Alternative Fuel Vehicle Eligible            75.0   

   Base MSRP  Legislative District  DOL Vehicle ID  \
0        0.0                  15.0       347724772   
1        0.0                  35.0       272165288   
2        0.0                  32.0       203182584   
3        0.0                  30.0       349010287   
4        0.0                  40.0       131684150   

              Vehicle Location                               Electric Utility  \
0  POINT (-120.50729 46.60464)                                     PACIFICORP   
1  POINT (-122.68471 47.50524)                         PUGET SOUND ENERGY INC   
2  POINT (-122.29245 47.82557)                         PUGET SOUND ENERGY INC   
3   POINT (-122.23035 47.3074)  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)   
4  POINT (-122.33891 48.41644)                         PUGET SOUND ENERGY INC   

   2020 Census Tract  
0       5.307700e+10  
1       5.303509e+10  
2       5.306105e+10  
3       5.303303e+10  
4       5.305795e+10

In [3]:
df.shape #show the shape of the data set

(257635, 17)

In [8]:
df.info() #Gives detailed info about what the data container has

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 257635 entries, 0 to 257634
Data columns (total 17 columns):
 #   Column                                             Non-Null Count   Dtype  
---  ------                                             --------------   -----  
 0   VIN (1-10)                                         257635 non-null  object 
 1   County                                             257627 non-null  object 
 2   City                                               257627 non-null  object 
 3   State                                              257635 non-null  object 
 4   Postal Code                                        257627 non-null  float64
 5   Model Year                                         257635 non-null  int64  
 6   Make                                               257635 non-null  object 
 7   Model                                              257635 non-null  object 
 8   Electric Vehicle Type                              257635 non-null  object

In [7]:
df.describe #Gives basic statistics about the data set overall

Postal Code     Model Year  Electric Range      Base MSRP  \
count  257627.000000  257635.000000   257632.000000  257632.000000   
mean    98177.614217    2021.713164       43.127775     705.270541   
std      2535.922185       3.023857       81.632805    6997.364357   
min      1469.000000    2000.000000        0.000000       0.000000   
25%     98052.000000    2020.000000        0.000000       0.000000   
50%     98133.000000    2023.000000        0.000000       0.000000   
75%     98380.000000    2024.000000       35.000000       0.000000   
max     99577.000000    2026.000000      337.000000  845000.000000   

       Legislative District  DOL Vehicle ID  2020 Census Tract  
count         257038.000000    2.576350e+05       2.576270e+05  
mean              28.897151    2.404009e+08       5.297530e+10  
std               14.882168    6.602927e+07       1.597188e+09  
min                1.000000    4.385000e+03       1.001020e+09  
25%               17.000000    2.134322e+08       5.303301e+10  
50%               32.000000    2.586345e+08       5.303303e+10  
75%               42.000000    2.738272e+08       5.305307e+10  
max               49.000000    4.792548e+08       6.601095e+10

## Evaluation/First Impressions

This data set consists of 17 columns and 257,635 rows, making it extremely large. The columns conist of manily qualitative data, but there are a couple quatitative columns. Overall, this is a very nicely organizted data set, with probably not a lot of missing entries.

# Cleaning the Data

In [15]:
df.isnull().sum() #Check how many missing cells there are

VIN (1-10)                                             0
County                                                 8
City                                                   8
State                                                  0
Postal Code                                            8
Model Year                                             0
Make                                                   0
Model                                                  0
Electric Vehicle Type                                  0
Clean Alternative Fuel Vehicle (CAFV) Eligibility      0
Electric Range                                         3
Base MSRP                                              3
Legislative District                                 597
DOL Vehicle ID                                         0
Vehicle Location                                      16
Electric Utility                                       8
2020 Census Tract                                      8
dtype: int64

In [19]:
df = df.dropna()
df.isnull().sum()

VIN (1-10)                                           0
County                                               0
City                                                 0
State                                                0
Postal Code                                          0
Model Year                                           0
Make                                                 0
Model                                                0
Electric Vehicle Type                                0
Clean Alternative Fuel Vehicle (CAFV) Eligibility    0
Electric Range                                       0
Base MSRP                                            0
Legislative District                                 0
DOL Vehicle ID                                       0
Vehicle Location                                     0
Electric Utility                                     0
2020 Census Tract                                    0
dtype: int64

# Finished Data Cleaning
Using the df.dropna() function is removed any of the missing values from the dataset, so there would not be any complications. I decieded not to rename any of the columns since they were all nicely organized and named already.

In [28]:
df['Model Year'] #Selected column by Model Year

0         2019
1         2024
2         2022
3         2017
4         2013
          ... 
257630    2020
257631    2022
257632    2019
257633    2019
257634    2025
Name: Model Year, Length: 257027, dtype: int64

In [22]:
df.sort_values(by='Model') #Sort data by model

VIN (1-10)   County        City State  Postal Code  Model Year  \
95471   3MW5P9J07M     King     Redmond    WA      98052.0        2021   
218647  WBA8E1C57G  Whatcom    Ferndale    WA      98248.0        2016   
25220   WBA8E1C35H    Mason     Belfair    WA      98528.0        2017   
121211  WBA8E1C57J     King     Seattle    WA      98119.0        2018   
72220   WBA5P9C02N    Clark  Ridgefield    WA      98642.0        2022   
...            ...      ...         ...   ...          ...         ...   
118231  2G5ZJ3HG8P     King      Seatac    WA      98148.0        2023   
114776  2G5ZJ3HG7P     King      Seatac    WA      98148.0        2023   
112118  2G5ZJ3HG2P     King      Seatac    WA      98148.0        2023   
103469  2G5ZJ3TZ9R     King      Auburn    WA      98002.0        2024   
117532  2G5ZJ3HG6P     King      Seatac    WA      98148.0        2023   

              Make Model                   Electric Vehicle Type  \
95471          BMW  330E  Plug-in Hybrid Electric Vehicle (PHEV)   
218647         BMW  330E  Plug-in Hybrid Electric Vehicle (PHEV)   
25220          BMW  330E  Plug-in Hybrid Electric Vehicle (PHEV)   
121211         BMW  330E  Plug-in Hybrid Electric Vehicle (PHEV)   
72220          BMW  330E  Plug-in Hybrid Electric Vehicle (PHEV)   
...            ...   ...                                     ...   
118231  BRIGHTDROP  ZEVO          Battery Electric Vehicle (BEV)   
114776  BRIGHTDROP  ZEVO          Battery Electric Vehicle (BEV)   
112118  BRIGHTDROP  ZEVO          Battery Electric Vehicle (BEV)   
103469  BRIGHTDROP  ZEVO          Battery Electric Vehicle (BEV)   
117532  BRIGHTDROP  ZEVO          Battery Electric Vehicle (BEV)   

        Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
95471               Not eligible due to low battery range            20.0   
218647              Not eligible due to low battery range            14.0   
25220               Not eligible due to low battery range            14.0   
121211              Not eligible due to low battery range            14.0   
72220               Not eligible due to low battery range            22.0   
...                                                   ...             ...   
118231  Eligibility unknown as battery range has not b...             0.0   
114776  Eligibility unknown as battery range has not b...             0.0   
112118  Eligibility unknown as battery range has not b...             0.0   
103469  Eligibility unknown as battery range has not b...             0.0   
117532  Eligibility unknown as battery range has not b...             0.0   

        Base MSRP  Legislative District  DOL Vehicle ID  \
95471         0.0                  45.0       144516550   
218647    43700.0                  42.0       128781418   
25220     44100.0                  35.0       250417023   
121211    45600.0                  36.0       104553844   
72220         0.0                  18.0       239731257   
...           ...                   ...             ...   
118231        0.0                  33.0       281868506   
114776        0.0                  33.0       282678973   
112118        0.0                  33.0       281812954   
103469        0.0                  47.0       275027602   
117532        0.0                  33.0       281849242   

                   Vehicle Location  \
95471   POINT (-122.13158 47.67858)   
218647   POINT (-122.5945 48.84787)   
25220   POINT (-122.82802 47.44998)   
121211   POINT (-122.3684 47.64586)   
72220   POINT (-122.74595 45.81539)   
...                             ...   
118231  POINT (-122.32863 47.46233)   
114776  POINT (-122.32863 47.46233)   
112118  POINT (-122.32863 47.46233)   
103469   POINT (-122.2217 47.30661)   
117532  POINT (-122.32863 47.46233)   

                                         Electric Utility  2020 Census Tract  
95471       PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303303e+10  
218647  PUGET SOUND ENERGY INC||PUD NO 1 

In [23]:
df[0:5] #Slice rows

VIN (1-10)     County          City State  Postal Code  Model Year    Make  \
0  5YJ3E1EB5K     Yakima        Yakima    WA      98901.0        2019   TESLA   
1  1C4RJXU67R     Kitsap  Port Orchard    WA      98367.0        2024    JEEP   
2  KNDCD3LD0N  Snohomish      Lynnwood    WA      98036.0        2022     KIA   
3  5UXKT0C37H       King        Auburn    WA      98001.0        2017     BMW   
4  1N4AZ0CP1D     Skagit  Mount Vernon    WA      98273.0        2013  NISSAN   

      Model                   Electric Vehicle Type  \
0   MODEL 3          Battery Electric Vehicle (BEV)   
1  WRANGLER  Plug-in Hybrid Electric Vehicle (PHEV)   
2      NIRO  Plug-in Hybrid Electric Vehicle (PHEV)   
3        X5  Plug-in Hybrid Electric Vehicle (PHEV)   
4      LEAF          Battery Electric Vehicle (BEV)   

  Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0           Clean Alternative Fuel Vehicle Eligible           220.0   
1             Not eligible due to low battery range            21.0   
2             Not eligible due to low battery range            26.0   
3             Not eligible due to low battery range            14.0   
4           Clean Alternative Fuel Vehicle Eligible            75.0   

   Base MSRP  Legislative District  DOL Vehicle ID  \
0        0.0                  15.0       347724772   
1        0.0                  35.0       272165288   
2        0.0                  32.0       203182584   
3        0.0                  30.0       349010287   
4        0.0                  40.0       131684150   

              Vehicle Location                               Electric Utility  \
0  POINT (-120.50729 46.60464)                                     PACIFICORP   
1  POINT (-122.68471 47.50524)                         PUGET SOUND ENERGY INC   
2  POINT (-122.29245 47.82557)                         PUGET SOUND ENERGY INC   
3   POINT (-122.23035 47.3074)  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)   
4  POINT (-122.33891 48.41644)                         PUGET SOUND ENERGY INC   

   2020 Census Tract  
0       5.307700e+10  
1       5.303509e+10  
2       5.306105e+10  
3       5.303303e+10  
4       5.305795e+10

In [27]:
df[df['Electric Range']>20] #Sorted by condition based on Electric Range over 20 miles

VIN (1-10)     County          City State  Postal Code  Model Year  \
0       5YJ3E1EB5K     Yakima        Yakima    WA      98901.0        2019   
1       1C4RJXU67R     Kitsap  Port Orchard    WA      98367.0        2024   
2       KNDCD3LD0N  Snohomish      Lynnwood    WA      98036.0        2022   
4       1N4AZ0CP1D     Skagit  Mount Vernon    WA      98273.0        2013   
5       1N4BZ0CP8G  Snohomish    Marysville    WA      98270.0        2016   
...            ...        ...           ...   ...          ...         ...   
257624  5YJ3E1EB4J    Whatcom    Bellingham    WA      98229.0        2018   
257625  WBY8P2C52K  Snohomish       Everett    WA      98208.0        2019   
257626  5YJ3E1EB5L    Spokane       Spokane    WA      99212.0        2020   
257630  2C4RC1L73L  Snohomish      Mukilteo    WA      98275.0        2020   
257633  JTDKARFP9K    Whatcom    Bellingham    WA      98229.0        2019   

            Make               Model                   Electric Vehicle Type  \
0          TESLA             MODEL 3          Battery Electric Vehicle (BEV)   
1           JEEP            WRANGLER  Plug-in Hybrid Electric Vehicle (PHEV)   
2            KIA                NIRO  Plug-in Hybrid Electric Vehicle (PHEV)   
4         NISSAN                LEAF          Battery Electric Vehicle (BEV)   
5         NISSAN                LEAF          Battery Electric Vehicle (BEV)   
...          ...                 ...                                     ...   
257624     TESLA             MODEL 3          Battery Electric Vehicle (BEV)   
257625       BMW                  I3          Battery Electric Vehicle (BEV)   
257626     TESLA             MODEL 3          Battery Electric Vehicle (BEV)   
257630  CHRYSLER            PACIFICA  Plug-in Hybrid Electric Vehicle (PHEV)   
257633    TOYOTA  PRIUS PRIME (PHEV)  Plug-in Hybrid Electric Vehicle (PHEV)   

       Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0                Clean Alternative Fuel Vehicle Eligible           220.0   
1                  Not eligible due to low battery range            21.0   
2                  Not eligible due to low battery range            26.0   
4                Clean Alternative Fuel Vehicle Eligible            75.0   
5                Clean Alternative Fuel Vehicle Eligible            84.0   
...                                                  ...             ...   
257624           Clean Alternative Fuel Vehicle Eligible           215.0   
257625           Clean Alternative Fuel Vehicle Eligible           153.0   
257626           Clean Alternative Fuel Vehicle Eligible           322.0   
257630           Clean Alternative Fuel Vehicle Eligible            32.0   
257633             Not eligible due to low battery range            25.0   

        Base MSRP  Legislative District  DOL Vehicle ID  \
0             0.0                  15.0       347724772   
1             0.0                  35.0       272165288   
2             0.0                  32.0       203182584   
4             0.0                  40.0       131684150   
5             0.0                  44.0       256278109   
...           ...                   ...             ...   
257624        0.0                  40.0       473977087   
257625        0.0                  44.0       260212564   
257626        0.0                   4.0         4783222   
257630        0.0                  21.0       161576068   
257633        0.0                  40.0       177812232   

                   Vehicle Location  \
0       POINT (-120.50729 46.60464)   
1       POINT (-122.68471 47.50524)   
2       POINT (-122.29245 47.82557)   
4       POINT (-122.33891 48.41644)   
5        POINT (-122.1769 48.06114)   
...                             ...   
257624   POINT (-122.45486 48.7449)   
257625  POINT (-122.18637 47.89251)   
257626  POINT (-117.28805 47.68043)   
257630  POINT (-122.29196 47.89908)   
257633   POINT (-122.45486 48.7449)   

                   